sugondat-shim/src/cli.rs

```diff
@@ -153,64 +148,104 @@ pub mod query {
         Submit(submit::Params),
         /// Queries information about a block and header.
         Block(block::Params),
+        /// Queries information about a specific blob.
+        Blob(blob::Params),
     }
 
-    pub mod block {
-        //! CLI definition for the `query block` subcommand.
+    /// A reference to a block to query.
+    #[derive(Debug, Clone)]
+    pub enum BlockRef {
+        /// The current best finalized block known by the node.
+        Best,
+        /// The number of the block to query.
+        Number(u64),
+        /// The hex-encoded hash of the block to query, prefixed with "0x".
+        Hash([u8; 32]),
+    }
 
-        use clap::Args;
+    impl std::fmt::Display for BlockRef {
+        fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+            match *self {
+                BlockRef::Best => write!(f, "best"),
+                BlockRef::Number(n) => write!(f, "{}", n),
+                BlockRef::Hash(h) => write!(f, "0x{}", hex::encode(&h[..])),
+            }
+        }
+    }
+
+    impl std::str::FromStr for BlockRef {
+        type Err = String;
+
+        fn from_str(input: &str) -> Result<Self, Self::Err> {
+            if input == "best" {
+                return Ok(BlockRef::Best);
+            }
+
+            if let Some(hash) = decode_hash(input)? {
+                return Ok(BlockRef::Hash(hash));
+            }
 
-        use super::SugondatRpcParams;
-
-        /// A reference to a block to query.
-        #[derive(Debug, Clone)]
-        pub enum BlockRef {
-            /// The current best finalized block known by the node.
-            Best,
-            /// The number of the block to query.
-            Number(u64),
-            /// The hex-encoded hash of the block to query, prefixed with "0x".
-            Hash([u8; 32]),
+            if let Ok(n) = input.parse::<u64>() {
+                Ok(BlockRef::Number(n))
+            } else {
+                Err(format!("parse error. see `--help`"))
+            }
         }
+    }
 
-        impl std::fmt::Display for BlockRef {
-            fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
-                match *self {
-                    BlockRef::Best => write!(f, "best"),
-                    BlockRef::Number(n) => write!(f, "{}", n),
-                    BlockRef::Hash(h) => write!(f, "0x{}", hex::encode(&h[..])),
-                }
+    fn decode_hash(input: &str) -> Result<Option<[u8; 32]>, String> {
```

Comment:
Nit: Would appreciate not mixing new features and refactorings

Reply:
I am not graphite-pilled enough to keep things that small. at the time, it would've required stacking on top of only some of the changes within the file, stashing the other ones, and then popping the stash to build on top of the refactored decoder

Reply:
Fair enough. I suggested that because for me it seemed like a simple operation.